src/rewrite/vct/rewrite/lang/LangLLVMToCol.scala

```diff
@@ -2,57 +2,305 @@ package vct.rewrite.lang
+        case (TPointer(a), TPointer(b)) => moreSpecific(a, b)
+        case (LLVMTStruct(_, _, a), LLVMTStruct(_, _, b)) =>
+          a.headOption.exists(ta => b.exists(tb => moreSpecific(ta, tb)))
+        case (LLVMTStruct(_, _, _), _) => true
+        case (LLVMTArray(_, a), LLVMTArray(_, b)) => moreSpecific(a, b)
+        case (LLVMTArray(_, _), _) => true
+        case _ => false
+      }
+    }
+
+    // TODO: This sorting is non-stable which might cause nondeterministic bugs if there's something wrong with moreSpecific
+    def findMostSpecific(
+        types: mutable.ArrayBuffer[Type[Pre]]
+    ): Option[Type[Pre]] = {
+      types.map(Some(_)).reduce[Option[Type[Pre]]] { (a, b) =>
+        (a, b) match {
+          case (None, _) | (_, None) => None
+          case (Some(a), Some(b))
+              if a == b || rw.dispatch(a) == rw.dispatch(b) ||
+                moreSpecific(a, b) =>
+            Some(a)
+          case (Some(a), Some(b)) if moreSpecific(b, a) => Some(b)
+          case _ => None
+        }
+      }
+    }
+
+    class TypeGuess(
+        val depends: mutable.Set[Object] = mutable.Set(),
+        val dependents: mutable.Set[Object] = mutable.Set(),
+        val getGuesses: mutable.ArrayBuffer[Unit => Type[Pre]] = mutable
+          .ArrayBuffer(),
+        var currentType: Type[Pre],
+    ) {
+      def add(dependencies: Set[Object], inferType: Unit => Type[Pre]): Unit = {
+        depends.addAll(dependencies)
+        getGuesses.addOne(inferType)
+      }
+
+      def update(): Boolean = {
+        val superType = findMostSpecific(getGuesses.map(_()))
+        if (superType.isEmpty) { false }
+        else {
+          val updated = currentType == superType.get
+          currentType = superType.get
+          updated
+        }
+      }
+    }
+
+    val typeGuesses: mutable.HashMap[Object, TypeGuess] = mutable.HashMap()
+
+    def findDependencies(expr: Expr[Pre]): Set[Object] = {
+      expr.collect {
+        case Local(Ref(v)) => v
+        case LLVMPointerValue(Ref(g)) => g
+        // These two below probably don't do anything
+        case v: Variable[Pre] => v
+        case v: LLVMGlobalVariable[Pre] => v
+      }.toSet
+    }
+
+    def replaceWithGuesses(
+        value: Expr[Pre],
+        dependencies: Set[Object],
+    ): Expr[Pre] = {
+      val subMap = dependencies.filter(typeGuesses.contains).collect {
+        case v: Variable[Pre] if typeGuesses(v).currentType != v.t =>
+          (v, new Variable[Pre](typeGuesses(v).currentType)(v.o))
+        case v: LLVMGlobalVariable[Pre]
+            if typeGuesses(v).currentType != v.variableType =>
+          (
+            v,
+            new LLVMGlobalVariable[Pre](
+              typeGuesses(v).currentType,
+              v.value,
+              v.constant,
+            )(v.o),
+          )
+      }
+      if (subMap.isEmpty) { value }
+      else {
+        // TODO: Support multiple guesses?
```

Comment:
Only if there's an actual use case, I don't imagine any sane compiler does this (I'm probably wrong! :D)